### PR TITLE
 Prisoner availability API

### PR DESCRIPTION
**Pull request by StupidCodeFactory**
**CHANGES:**

Create an API to provide a prisoner's availability

### CODE REVIEW

File: app/services/prisoner_slot_availability.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  def slots
+    if enforce_all_available_slots?
+      return all_slots_available
+    end
+
+    results = Hash.new { |h, slot| h[slot] = [] }
+    prison_slots.each_with_object(results) do |slot, slots_with_availabilities|
+      slots_with_availabilities[slot.to_s]
```
Comment (alan) [inline app/services/prisoner_slot_availability.rb]: Why read the value but do nothing with it?

File: config/environments/test.rb [hunk 2]
```diff
@@ -21,7 +21,7 @@
   # Always make testing calls to the NOMIS dev server, and hard-code
   # credentials to avoid the possibility of leaking sensitive credentials or
   # returned data into VCR test output files.
-  config.nomis_api_host = 'http://172.22.16.2:8080/'
+  config.nomis_api_host = 'http://151.237.239.116:4888'
```
Comment (alan) [inline config/environments/test.rb]: Why the change of ip?

File: app/services/prisoner_slot_availability.rb (hunk 1 above)
Comment (StupidCodeFactory) [inline app/services/prisoner_slot_availability.rb]: That's creating a the key to the ```results``` hash

File: config/environments/test.rb (hunk 2 above)
Comment (StupidCodeFactory) [inline config/environments/test.rb]: Because it's the new IP, the other one wasn't working

File: app/services/prisoner_slot_availability.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    self.api_error     = false
+  end
+
+  def slots
+    if enforce_all_available_slots?
+      return all_slots_available
+    end
+
+    all_slots_available.each do |slot, unavailability_reasons|
+      unless offender_availabilities_dates.delete(slot.to_date)
```
Comment (alan) [inline app/services/prisoner_slot_availability.rb]: I think this might have a bug. If there are 2 slots for the same date, then only one of the slots will be marked as 'prisoner unavailable' ?

I'd prefer if this method didn't have side effects, so that it would calculate the overall slot availability (`all_slots_available `) each time and also not modify the `offender_availabilities` list.

File: app/services/prisoner_slot_availability.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def enforce_all_available_slots?
+    !nomis_public_prisoner_availability_enabled? ||
+      !offender.valid? ||
+      (load_offender_availabilities && api_error)
+  end
+
+  def nomis_public_prisoner_availability_enabled?
+    Rails.configuration.nomis_public_prisoner_availability_enabled
```
Comment (alan) [inline app/services/prisoner_slot_availability.rb]: We probably also want to check `Nomis::Api.enabled?` here

File: config/environments/test.rb (hunk 2 above)
Comment (alan) [inline config/environments/test.rb]: Is that the dev environment?

File: spec/controllers/api/slots_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        prisoner_number: prisoner.number,
+        date_of_birth:   prisoner.date_of_birth,
+        start_date:      '2016-02-15',
+        end_date:        '2016-04-15'
+      }
+    }
+
+    let(:slots) {
+      [
+        { '2016-02-15T13:30/14:30' => true },
```
Comment (alan) [inline spec/controllers/api/slots_controller_spec.rb]: I guess the calendar doesn't need the specific reason is not available?

File: spec/services/prisoner_slot_availability_spec.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  let(:start_date)    { Date.parse('2017-02-10') }
+  let(:end_date)      { Date.parse('2017-04-11') }
+
+  subject do
+    described_class.new(prison, offender_id, date_of_birth, start_date..end_date)
+  end
+
+  def all_slots_available
+    {
+      ConcreteSlot.new(2017, 2, 13, 14, 0, 16, 10).to_s => [],
```
Comment (alan) [inline spec/services/prisoner_slot_availability_spec.rb]: Can we make the test more concise by removing some of the slots? I don't think we need so many slots to test this class

File: config/environments/test.rb (hunk 2 above)
Comment (StupidCodeFactory) [inline config/environments/test.rb]: yes

File: app/services/prisoner_slot_availability.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
+class PrisonerSlotAvailability
+  PRISONER_UNAVAILABLE = 'prisoner_unavailable'.freeze
```
Comment (alan) [inline app/services/prisoner_slot_availability.rb]: I'm pretty sure we have the same string defined as a constant somewhere else. We should think of a good way of sharing constants to avoid duplicates thorough the app. Just a general comment, not meant to be addressed in this PR.

File: app/services/prisoner_slot_availability.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    self.prison        = prison
+    self.noms_id       = noms_id
+    self.date_of_birth = date_of_birth
+    self.start_date    = date_range.min
+    self.end_date      = date_range.max
+    self.api_error     = false
+  end
+
+  def slots
+    if enforce_all_available_slots?
```
Comment (alan) [inline app/services/prisoner_slot_availability.rb]: Inline the `if`?

File: app/services/prisoner_slot_availability.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      return all_slots
+    end
+
+    results = all_slots.deep_dup.each { |slot, unavailability_reasons|
```
Comment (alan) [inline app/services/prisoner_slot_availability.rb]: 👍

**Comment by alan** [comment]
@StupidCodeFactory not sure if you saw my previous comment about validating the dates, however after seeing the change in public, perhaps we can remove the dates?
